code/weapon/beam.cpp

```diff
@@ -2152,6 +2224,229 @@ void beam_get_binfo(beam *b, float accuracy, int num_shots)
+			usable_target = b->target;
+
+		// set up shooter orient now
+		matrix orient = vmd_identity_matrix;
+		if (b->flags & BF_IS_FIGHTER_BEAM) {
+			orient = b->objp->orient;
+		} else if (b->subsys) {
+			vec3d fvec, uvec;
+			vm_vec_sub(&fvec, &b->target->pos, &turret_point);
+			vm_vec_unrotate(&uvec, &b->subsys->system_info->turret_norm, &b->objp->orient);
+			vm_vector_2_matrix(&orient, &fvec, &uvec);
+		} else if (b->flags & BF_TARGETING_COORDS) {
+			// targeting coords already set up turret_norm with target_pos above
+			vm_vector_2_matrix(&orient, &turret_norm);
+		} 
+
+		vec3d rand1_on, rand2_on;      vm_vec_zero(&rand1_on);  vm_vec_zero(&rand2_on);
+		vec3d rand1_off, rand2_off;    vm_vec_zero(&rand1_off); vm_vec_zero(&rand2_off);
+
+		// Get our two starting points
+		if (usable_target) {
+			// set up our two kinds of random points if needed
+			if (bwi->bpi.start_pos == POS_RANDOM_INSIDE || bwi->bpi.end_pos == POS_RANDOM_INSIDE) {
+				vec3d temp1, temp2;
+				submodel_get_two_random_points_better(model_num, 0, &temp1, &temp2, seed);
+				vm_vec_rotate(&rand1_on, &temp1, &b->target->orient);
+				vm_vec_rotate(&rand2_on, &temp2, &b->target->orient);
+				rand1_on += b->target->pos;
+				rand2_on += b->target->pos;
+			}
+			if (bwi->bpi.start_pos == POS_RANDOM_OUTSIDE || bwi->bpi.end_pos == POS_RANDOM_OUTSIDE)
+				beam_get_octant_points(model_num, usable_target, seed % BEAM_NUM_GOOD_OCTANTS, Beam_good_slash_octants, &rand1_off, &rand2_off);
+
+			// get start and end points
+			switch (bwi->bpi.start_pos) {
+				case POS_CENTER:
+					pos1 = b->target->pos;
+					break;
+				case POS_RANDOM_INSIDE:
+					pos1 = rand1_on;
+					break;
+				case POS_RANDOM_OUTSIDE:
+					pos1 = rand1_off;
+			}
+
+
+			if (bwi->bpi.no_translate || bwi->bpi.end_pos == POS_SAME_RANDOM)
+				pos2 = pos1;
+			else {
+				switch (bwi->bpi.end_pos) {
+					case POS_CENTER:
+						pos2 = b->target->pos;
+						break;
+					case POS_RANDOM_INSIDE:
+						pos2 = rand2_on;
+						break;
+					case POS_RANDOM_OUTSIDE:
+						pos2 = rand2_off;
+				}
+			}
+
+			// set rot_axis if its center
+			if (bwi->bpi.continuous_rot_axis == AXIS_CENTER) 
+				rot_axis = b->target->pos;
+			if (bwi->bpi.per_burst_rot_axis == AXIS_CENTER)
+				per_burst_rot_axis = b->target->pos;
+			if (bwi->bpi.burst_rot_axis == AXIS_CENTER)
+				burst_rot_axis = b->target->pos;
+			
+		} else { 
+			vec3d center;
+			vm_vec_zero(&center);
+			// if we have no target let's act as though we're shooting at something with a 300m radius 800m away
+			if (bwi->bpi.start_pos != POS_CENTER)
+				static_rand_cone(rand32(), &pos1, &center, 0.f, 180.f);
+
+			if (bwi->bpi.end_pos != POS_CENTER)
+				static_rand_cone(rand32(), &pos1, &center, 0.f, 180.f);
+
+			if (bwi->bpi.no_translate || bwi->bpi.end_pos == POS_SAME_RANDOM)
+				pos2 = pos1;
+
+			pos1 *= 300.f;
+			pos2 *= 300.f;
+			vec3d move_forward;
+			vm_vec_make(&move_forward, 0.f, 0.f, 800.f);
```

Comment:
> // if we have no target let's act as though we're shooting at something with a 300m radius 800m away

Due to the way the beams are configured I think it will lead to a lot less user confusion down the like if the distance and radius values of the 'virtual target' in these cases are equal